provd/internal/services/user/user.go

```diff
@@ -186,51 +201,83 @@ func (s *Service) ValidateUsername(ctx context.Context, req *pb.ValidateUsername
-			continue
-		}
-		if line == username {
-			isReserved = true
-			break
-		}
+	// Open the group.master file
+	groupFile, err := os.Open("/usr/share/base-passwd/group.master")
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error opening group.master file: %v", err)
+	}
+	defer groupFile.Close()
+
+	// Check the passwd.master file
+	isReserved, err := scanFile(passwdFile, username)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "error reading passwd.master file: %v", err)
 	}
 
-	if err := scanner.Err(); err != nil {
-		return nil, status.Errorf(codes.Internal, "error reading reserved usernames file: %v", err)
+	// If not found, check the group.master file
+	if !isReserved {
+		isReserved, err = scanFile(groupFile, username)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "error reading group.master file: %v", err)
+		}
 	}
 
+	// If found, return reserved
 	if isReserved {
 		return &pb.ValidateUsernameResponse{UsernameValidation: pb.UsernameValidation_SYSTEM_RESERVED}, nil
 	}
 
 	// Check if username is already in use
-	err = s.Accounts.Call(consts.DbusAccountsPrefix+".FindUserByName", 0, username).Err
-	if err != nil {
-		var dbusError dbus.Error
-		if errors.As(err, &dbusError) {
-			if dbusError.Name == consts.DbusAccountsPrefix+".Error.Failed" {
-				// User not found
-				return &pb.ValidateUsernameResponse{UsernameValidation: pb.UsernameValidation_OK}, nil
-			}
-			// Handle other dbus errors
-			return nil, status.Errorf(codes.Internal, "dbus error: %v", dbusError)
-		}
-		// Unknown error
+	err = s.accounts.Call(consts.DbusAccountsPrefix+".FindUserByName", 0, username).Err
+
+	// User found
+	if err == nil {
+		return &pb.ValidateUsernameResponse{UsernameValidation: pb.UsernameValidation_ALREADY_IN_USE}, nil
+	}
+
+	// Inspect the error we got
+	var dbusError dbus.Error
+	isDbusError := errors.As(err, &dbusError)
+
+	// Unknown error
+	if !isDbusError {
 		return nil, status.Errorf(codes.Internal, "unknown error: %v", err)
 	}
 
-	// User found
-	return &pb.ValidateUsernameResponse{UsernameValidation: pb.UsernameValidation_ALREADY_IN_USE}, nil
+	// Non "user not found" account service error
+	errBody, ok := dbusError.Body[0].(string)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "unknown error: %v", err)
+	}
+
+	if dbusError.Name != consts.DbusAccountsPrefix+".Error.Failed" || !strings.Contains(errBody, username) {
+		return nil, status.Errorf(codes.Internal, "dbus error: %v", dbusError)
+	}
+
+	// We have "user not found"
+	return &pb.ValidateUsernameResponse{UsernameValidation: pb.UsernameValidation_OK}, nil
+}
+
+// Function to scan a file for the username.
+func scanFile(file *os.File, username string) (bool, error) {
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		if strings.HasPrefix(line, "#") || line == "" {
+			continue
+		}
+		// Extract the username from the line
+		fields := strings.SplitN(line, ":", 2)
```

Comment:
I would error out if the file does not have the expected format, So failing early, continuing if not found, and the "happy path" is one tab more left, with the return.